### DISCUSSION TITLE
Cannot create contact with same email as an existing host

### DISCUSSION THREAD

**Discussion by c1wei** [iotum/community#21] [Ideas]
My host (HostA) has another coworker host (HostB) that has email "email_redacted@email.com"

If my HostA tries to add a contact with "email_redacted@email.com" this will not create a new contact, and instead respond with the HostB data. This is being done via contact/create Iotum API endpoint.

**Comment by noamtiotum** [comment]
@c1wei we can't allow the email of a host/contact to be edited. It doesn't make sense on our platform, it would be like changing the extension of a another user on your platform.  Similar to how in Netspaiens you can't change the contact's name if they are a user.

We are going to figure out a way to allow users to add other contact details like adding phone numbers.

**Comment by noamtiotum** [comment]
# Background

The Iotum platform has always made the email address of a Host unique within the scope of a Company.  While the platform has the concept of username and could potentially allow multiple hosts to have the same email address we have not because we've always felt if could cause confusion.  The same goes for creating contacts with the same email address as a Host, we don't allow it because we feel it can get confusing.

Specifically, it could be confusing when a Host is scheduling a meeting.  Email is the default channel for scheduling, so if you were presented with multiple options with the same email it would be hard to know which you should choose.  The Host scheduling may also then not feel confident they have invited the person correctly.

For the person being invited it could also be confusing. For example, if the scheduler had a contact with an email and a Host with the same email and they invited the contact, the person that owns that email would get the invitation but when they log into their account they would not see the invitation.  Another example is if there are two Hosts with the same email address and the scheduler invites one but the person that was invited mainly uses the other account they would not see the meeting invitation in their meeting dashboard resulting in customer service requests and confused users.

For platforms that use another main key for accounts like a username or extension this restriction on email address can cause some conflict.  Platforms like these will want to create Iotum Hosts for everyone one of their users but will get errors when trying to create a Host with an email address that already exists or creating a contact when an existing contact or Host already exists.

# Proposed Solutions for Iotum Hosts

## 1. Partners create alternative email addresses if an address is already taken

For example if `noamt@itoum.com` was taken, the partner would add a “-1” on the user part of the email like `noamt-1@iotum.com`.  This would remove any errors when creating Hosts.  The user would still be able to access all their accounts by using the partner applicable.

The downside would be that some users may end up with Host accounts with the wrong email.  This would be able to be fixed easily by a support agent in the Iotum Admin App.

If the number of real accounts that this happens to is a small minority this is a good solution.

## 2. Create only 1 Iotum Host account for every email in the partner application

For example, if multiple partner accounts had the email noamt@iotum.com, only create 1 Iotum Host account and link both partner accounts to that Iotum Host.  

The nice thing about this is that regardless of which partner account the user uses they will always end up in the same Iotum account with all their meeting visible.

Where this solution may have some edge cases is when the user changes their email in the partner's application in one of the accounts.  Now you have 2 partner accounts with different emails and a single Iotum account where the email only matches one of them.

While this is a bit awkward, it may not cause too many problems.

## 3. Iotum shows the username of the Iotum Host account when there are multiple accounts with the same email

For example, Iotum would allow the partner to create multiple Host accounts with the same email.  If multiple Hosts had the email `noamt@iotum.com`, Iotum would present this user as `noamt@iotum.com (1000@iotumtest)` and `noamt@iotum.com (1008@iotumtest)`.

Hosts trying to invite me to a meeting would see two distinct options to invite me to a meeting.  The partner would not get any errors in creating Hosts.

However the downside would still be that this could be confusing for the scheduler and the person invited.  The scheduler would have to pick between two accounts that are only differentiated by the username, the scheduler may not know which to use.  Additionally, the person invited could get invitations to meetings that may not appear in the account they use more often.

**Reply by noamtiotum** [reply]
We are going to attempt to implement Solution 3 and hope that works.

**Comment by noamtiotum** [comment]
@c1wei for the situation where you try to create or edit a contact via contact/create contact/update and the email address matches an existing host on the same company we are going to modify the API to return a 409 Conflict so you will know exactly why it's failed.

We'll give you more details when it's ready.
